File: android/src/main/kotlin/com/squareup/square_mobile_payments_sdk/SquareMobilePaymentsSdkPlugin.kt
package com.squareup.square_mobile_payments_sdk

import com.squareup.sdk.mobilepayments.MobilePaymentsSdk
import com.squareup.square_mobile_payments_sdk.modules.AuthModule
import com.squareup.square_mobile_payments_sdk.modules.PaymentModule
import com.squareup.square_mobile_payments_sdk.modules.ReaderModule
import com.squareup.square_mobile_payments_sdk.modules.SettingsModule
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.EventChannel.StreamHandler
import io.flutter.plugin.common.EventChannel.EventSink

/** SquareMobilePaymentsSdkPlugin */
class SquareMobilePaymentsSdkPlugin : FlutterPlugin, MethodCallHandler, StreamHandler {
  /// The MethodChannel that will the communication between Flutter and native Android
  ///
  /// This local reference serves to register the plugin with the Flutter Engine and unregister it
  /// when the Flutter Engine is detached from the Activity
  private lateinit var channel: MethodChannel
  private lateinit var eventChannel: EventChannel
  private var eventSink: EventSink? = null

  override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
    channel = MethodChannel(flutterPluginBinding.binaryMessenger, "square_mobile_payments_sdk")
    channel.setMethodCallHandler(this)

    eventChannel = EventChannel(flutterPluginBinding.binaryMessenger, "square_mobile_payments_sdk/events")
    eventChannel.setStreamHandler(this)
  }

  override fun onMethodCall(call: MethodCall, result: Result) {
    when (call.method) {
      "getPlatformVersion" ->
        result.success("Android Testing ${android.os.Build.VERSION.RELEASE}")

      "getSdkVersion" ->
        SettingsModule.getSdkVersion(result)

      "getEnvironment" ->
        SettingsModule.getEnvironment(result)

      "getAuthorizationState" ->
        AuthModule.getAuthorizationState(result)

      "getAuthorizedLocation" ->
        AuthModule.getAuthorizedLocation(result)

      "authorize" -> {
        val accessToken = call.argument<String>("accessToken")
        val locationId = call.argument<String>("locationId")
        AuthModule.authorize(result, accessToken, locationId)
      }
      "deauthorize" -> AuthModule.deAuthorize(result)
      "showMockReaderUI" -> ReaderModule.showMockReaderUI(result)
      "hideMockReaderUI" -> ReaderModule.hideMockReaderUI(result)
      "showSettings" -> SettingsModule.showSettings(result)
      "startPayment" -> {
        val paymentParameters = call.argument<HashMap<String, Any>>("paymentParameters")
        val promptParameters = call.argument<HashMap<String, Any>>("promptParameters")
        PaymentModule.startPayment(result, paymentParameters, promptParameters)
      }
      "isOfflineProcessingAllowed" -> SettingsModule.isOfflineProcessingAllowed(result)
      "getOfflineTotalStoredAmountLimit" -> SettingsModule.getOfflineTotalStoredAmountLimit(result)
      "getOfflineTransactionAmountLimit" -> SettingsModule.getOfflineTransactionAmountLimit(result)
      "getPayments" -> PaymentModule.getPayments(result)
      "getTotalStoredPaymentAmount" -> PaymentModule.getTotalStoredPaymentAmount(result)
      "getReaders" -> ReaderModule.getReaders(result)
      "getReader" -> {
        val id = call.argument<String>("id") as? String ?: ""
        ReaderModule.getReader(result, id)
      }
      "forget" -> {
        val id = call.argument<String>("id") as? String ?: ""
        ReaderModule.forget(result, id)
      }
      "blink" -> {
        val id = call.argument<String>("id") as? String ?: ""
        ReaderModule.blink(result, id)
      }
      "isPairingInProgress" -> ReaderModule.isPairingInProgress(result)
      "setReaderChangedCallback" -> ReaderModule.setReaderChangedCallback(result, eventSink)
      "removeReaderChangedCallback" -> ReaderModule.removeReaderChangedCallback(result)
      "pairReader" -> ReaderModule.pairReader(result)
      "stopPairing" -> ReaderModule.stopPairing(result)
      else -> result.notImplemented()
    }
  }

  // EventChannel handlers
  override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
    eventSink = events
  }

  override fun onCancel(arguments: Any?) {
    eventSink = null
  }

  override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
    channel.setMethodCallHandler(null)
    eventChannel.setStreamHandler(null)
  }
}

File: ios/Classes/SquareMobilePaymentsSdkPlugin.swift
import Flutter
import UIKit

public class SquareMobilePaymentsSdkPlugin: NSObject, FlutterPlugin {
  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(name: "square_mobile_payments_sdk", binaryMessenger: registrar.messenger())
    let instance = SquareMobilePaymentsSdkPlugin()
    registrar.addMethodCallDelegate(instance, channel: channel)
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "getPlatformVersion":
      result("iOS " + UIDevice.current.systemVersion)
    case "authorize":
      if  let arguments = call.arguments as? [String: Any],
          let accessToken = arguments["accessToken"] as? String,
          let locationId = arguments["locationId"] as? String {
        AuthModule.authorize(result: result, accessToken: accessToken, locationId: locationId)
      } else {
        result(FlutterError(
          code: "INVALID_ARGUMENTS",
          message: "Missing accessToken or locationId",
          details: nil))
      }
    case "deauthorize":
      AuthModule.deauthorize(result: result)
    case "getAuthorizationState":
      AuthModule.getAuthorizationState(result: result)
    case "showMockReaderUI":
      ReaderModule.showMockReaderUI(result: result)
    case "hideMockReaderUI":
      ReaderModule.hideMockReaderUI(result: result)
    case "linkAppleAccount":
      ReaderModule.linkAppleAccount(result: result)
    case "relinkAppleAccount":
      ReaderModule.relinkAppleAccount(result: result)
    case "isDeviceCapable":
      ReaderModule.isDeviceCapable(result: result)
    case "isAppleAccountLinked":
      ReaderModule.isAppleAccountLinked(result: result)
    case "startPayment":
      if  let arguments = call.arguments as? [String: Any],
          let paymentParameters = arguments["paymentParameters"] as? [String: Any],
          let promptParameters = arguments["promptParameters"] as? [String: Any] {
          PaymentModule.startPayment(
            result: result,
            paymentParameters: paymentParameters,
            promptParameters: promptParameters)
      } else {
        result(FlutterError(
            code: "missingParameters",
            message: "paymentParameters or promptParameters must not be null",
            details: nil))
      }
    case "showSettings":
      SettingsModule.showSettings(result: result)

    case "getSdkVersion":
      SettingsModule.getSdkVersion(result: result)

    case "getEnvironment":
      SettingsModule.getEnvironment(result: result)
    case "isOfflineProcessingAllowed":
      SettingsModule.isOfflineProcessingAllowed(result: result)
    case "getOfflineTotalStoredAmountLimit":
      SettingsModule.getOfflineTotalStoredAmountLimit(result: result)
    case "getOfflineTransactionAmountLimit":
      SettingsModule.getOfflineTransactionAmountLimit(result: result)
    case "getPayments":
      PaymentModule.getPayments(result: result)
    case "getTotalStoredPaymentAmount":
      PaymentModule.getTotalStoredPaymentAmount(result: result)

    default:
      result(FlutterMethodNotImplemented)
    }
  }
}
